db/background_mgr_dcp_client.go

```diff
@@ -0,0 +1,110 @@
+	ID                string                    // unique ID for this DCP client
+	FailOnRollback    bool                      // When true, the DCP client will terminate on DCP rollback
+	MetadataStoreType base.DCPMetadataStoreType // define storage type for DCPMetadata
+	GroupID           string                    // specify GroupID, only used when MetadataStoreType is DCPMetadataCS
+	CheckpointPrefix  string                    // Prefix used by
+	CollectionIDs     []uint32                  // CollectionIDs used by gocbcore, if empty, uses default collections
+	OneShot           bool                      // Whether the dcp feed will be continuous or not
+}
+
+// NewBackgroundManagerDcpClient create a new DCP client for background tasks. From the bucket you pass in it will either
+// create a rosmar dcp client or a gocb dcp client
+func NewBackgroundManagerDcpClient(ctx context.Context, bucket base.Bucket, options backgroundManagerDcpClientOptions) (backgroundMgrDcpClient, error) {
+	gocbBucket, err := base.AsGocbV2Bucket(bucket)
+	if err == nil {
+		clientOptions := &base.DCPClientOptions{
+			OneShot:           options.OneShot,
+			FailOnRollback:    options.FailOnRollback,
+			MetadataStoreType: options.MetadataStoreType,
+			GroupID:           options.GroupID,
+			CheckpointPrefix:  options.CheckpointPrefix,
+			CollectionIDs:     options.CollectionIDs,
+		}
+		return base.NewDCPClient(ctx, options.ID, options.Callback, *clientOptions, gocbBucket)
+	}
+	rosmarBucket, ok := base.GetBaseBucket(bucket).(*rosmar.Bucket)
+	if !ok {
+		return nil, fmt.Errorf("Invalid bucket type for background manager DCP client: %T", bucket)
+	}
+	return &rosmarDcpClient{
+		ctx:      ctx,
+		callback: options.Callback,
+		feedArgs: sgbucket.FeedArguments{
+			ID:         options.ID,
+			DoneChan:   make(chan struct{}),
+			Scopes:     options.Scopes,
+			Dump:       true,
+			Terminator: make(chan bool),
+		},
+		bucket:   rosmarBucket,
+		doneChan: make(chan error),
+	}, nil
+}
+
+type rosmarDcpClient struct {
+	ctx      context.Context
+	callback sgbucket.FeedEventCallbackFunc
+	feedArgs sgbucket.FeedArguments
+	bucket   *rosmar.Bucket
+	doneChan chan error
+}
+
+// Start returns an error and a channel to indicate when the DCPClient is done. If Start returns an error, DCPClient.Close() needs to be called.
+func (r *rosmarDcpClient) Start() (doneChan chan error, err error) {
+	err = r.bucket.StartDCPFeed(r.ctx, r.feedArgs, r.callback, nil)
+	go func() {
+		select {
+		case <-r.ctx.Done():
+			close(r.feedArgs.DoneChan)
+		case <-r.feedArgs.DoneChan:
+			close(r.doneChan)
+		}
+	}()
+	return r.doneChan, err
+}
+
+// Close is used externally to stop the DCP client. If the client was already closed due to error, returns that error
+func (r *rosmarDcpClient) Close() error {
+	close(r.feedArgs.Terminator)
+	return nil
+}
+
+// GetMetadata returns metadata for all vbuckets
+func (dc *rosmarDcpClient) GetMetadata() []base.DCPMetadata {
+	return nil
```

Comment:
I thought GetMetadata was used as part of stop/pause/resume for the gocb DCP feed.  If that's being done through MetadataStoreType, is GetMetadata needed as part of the API?

Reply:
It's used https://github.com/couchbase/sync_gateway/blob/d9a47e914ffa1dc7f995213339de2c4bb085bd52/db/background_mgr_resync_dcp.go#L183 but I do not think this is actually _used_ and it was copied from the attachment code (where it isn't in fact used). I have some evidence of this because `ResyncManagerDCP` literally calls `AttachmentManagerMeta` https://github.com/couchbase/sync_gateway/blob/d9a47e914ffa1dc7f995213339de2c4bb085bd52/db/background_mgr_resync_dcp.go#L350 and I don't think it _consumes_ this value at any point. Resuming resync works because it has a common set of checkpoints. We can't even do anything with only vbuuids.

While I do not think this value is used, I think we might have to write this value for backward compatibility. I haven't considered this.

Yet again, this might not be the right time to look at this, but there might never be a good time.

I think there's some value to putting this into 3.3 for increased testability, but that might imply 3.2.1 code, and this change might be bigger there. I do feel OK about no-oping this for rosmar for sure.